crates/derive/src/sources/calldata.rs

```diff
@@ -0,0 +1,93 @@
+//! CallData Source
+
+use crate::traits::{AsyncIterator, ChainProvider};
+use crate::types::BlockInfo;
+use crate::types::StageError;
+use crate::types::StageResult;
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use alloy_primitives::{Address, Bytes};
+use async_trait::async_trait;
+
+/// A data iterator that reads from calldata.
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub struct CalldataSource<CP>
+where
+    CP: ChainProvider + Send,
+{
+    /// The chain provider to use for the calldata source.
+    chain_provider: CP,
+    /// The address of the batcher contract.
+    batcher_address: Address,
+    /// Block Ref
+    block_ref: BlockInfo,
+    /// The L1 Signer.
+    signer: Address,
+    /// Current calldata.
+    calldata: Vec<Bytes>,
+    /// Whether the calldata source is open.
+    open: bool,
+}
+
+impl<CP: ChainProvider + Send> CalldataSource<CP> {
+    /// Creates a new calldata source.
+    pub fn new(
+        chain_provider: CP,
+        batcher_address: Address,
+        block_ref: BlockInfo,
+        signer: Address,
+    ) -> Self {
+        Self {
+            chain_provider,
+            batcher_address,
+            block_ref,
+            signer,
+            calldata: Vec::new(),
+            open: false,
+        }
+    }
+
+    /// Loads the calldata into the source if it is not open.
+    async fn load_calldata(&mut self) -> anyhow::Result<()> {
+        if self.open {
+            return Ok(());
+        }
+
+        let info = self
```

Comment:
Destructure the tuple?
```suggestion
        let (block_info, txs) = self
```